app.js

```diff
@@ -2,23 +2,93 @@ const express = require('express');
 const app = express();
 const ejs = require('ejs')
 const fs = require('fs');
+const path = require('path');
 const moment = require('moment')
 var cron = require('node-cron');
 
 app.set('view engine', 'ejs');
 app.use(express.static(__dirname + '/public'));
 
 
-const speedTest = require('@opstalent/speedtest.net');
+const testAvailable = require('check-internet-connected');
+const testSpeed = require('@opstalent/speedtest.net');
+
+let getFilepath = (date) => {
+    const path = './public/';
+    return path + date + ".json";
+};
+
+let getDates = () => {
+    const dir = './public/';
+    const files = fs.readdirSync(dir);
+    const dates = files.map((filename) => path.parse(filename).name);
+    return dates;
+}
+
+let getResults = (date) => {
+    const filepath = getFilepath(date);
+    const data = fs.readFileSync(filepath, 'utf-8');
+    return data.toString();
+};
+
+let writeDataPoint = (datapoint, key) => {
+    const date = moment().format("YYYY-MM-DD");
+    const filepath = getFilepath(date);
+    if (!fs.existsSync(filepath)) {
+        const json = {
+            "Speedtests": [],
+            "Availability": []
+        }
+        fs.writeFile(filepath, JSON.stringify(json), (err) => {
+            if (err) {
+                throw err;
+            }
+            console.log("JSON File Created: " + filepath)
+        });
+    }
+    fs.readFile(filepath, 'utf-8', (err, file) => {

```

Comment:
if I'm reading this correctly, this is synchronous code. Therefore it seems if the file doesn't exist, and we start writing it with writeFile, the `fs.readFile` call would run even if `fs.writeFile` has not yet completed. This will probably throw an error. I would suggest extracting the file update function and run it in two places: 1) the callback to writeFile 2) an "else" condition for `!fs.existsSync`

Reply:
Apologies for the late response, I haven't had time to look at this. I think the scenario you describe is an issue with this implementation. I will get round to sorting this when I have some time, but thank you for the catch!

Reply:
I opted to just change the `writeFile` to `writeFileSync` as it was the easiest thing to do. If its not performant feel free to make a followup change (I am not familiar with JS to comment on performance here).